includes/class-sensei-lesson.php

```diff
@@ -3822,6 +3822,54 @@ public function get_course_id( $lesson_id ) {
 
 	}
 
+	/**
+	 * Returns the course ID for a given set of lessons (with the key being the lesson ID)
+	 *
+	 * @since $$next-version$$
+	 * @access public
+	 *
+	 * @param array<int> $lesson_ids Array of Lesson IDs.
+	 * @return array<int, int|false> Array where the key is the lesson ID, and the value is the course ID (or false if not found)
+	 */
+	public function get_course_ids( $lesson_ids ) {
+		global $wpdb;
+
+		sort( $lesson_ids, SORT_NUMERIC );
+		$lesson_ids = array_unique( $lesson_ids, SORT_NUMERIC );
+
+		$cache_key     = 'lesson/get-course-ids/' . md5( implode( ',', $lesson_ids ) );
+		$cache_group   = 'sensei';
+		$cached_result = wp_cache_get( $cache_key, $cache_group );
+		if ( false !== $cached_result ) {
+			return $cached_result;
+		}
+		$courses_by_lesson = array_fill_keys( $lesson_ids, false );
+		$placeholders      = implode( ', ', array_fill( 0, count( $lesson_ids ), '%d' ) );
+
+		// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared,WordPress.DB.PreparedSQLPlaceholders.UnfinishedPrepare
+		$query = $wpdb->prepare(
+			"
+				SELECT lesson.ID AS lesson_id, course.ID AS course_id FROM {$wpdb->posts} lesson
+					INNER JOIN {$wpdb->postmeta} AS lesson_meta ON lesson_meta.post_id=lesson.ID AND lesson_meta.meta_key='_lesson_course'
+					INNER JOIN {$wpdb->posts} AS course ON course.ID = lesson_meta.meta_value AND course.post_type='course'
+					WHERE lesson.ID in ({$placeholders})
+				  		AND lesson.post_type='lesson'
+		",
+			$lesson_ids
+		);
+		// phpcs:enable WordPress.DB.PreparedSQL.InterpolatedNotPrepared,WordPress.DB.PreparedSQLPlaceholders.UnfinishedPrepare
+
+		// phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery,WordPress.DB.PreparedSQL.NotPrepared
+		$results = $wpdb->get_results( $query );
+		if ( is_array( $results ) ) {
+			foreach ( $results as $result ) {
+				$courses_by_lesson[ $result->lesson_id ] = $result->course_id;
+			}
+			wp_cache_set( $cache_key, $courses_by_lesson, $cache_group );
```

Comment:
One thing that I just thought about: Right now, I'm calling `wp_cache_set` with the default expiration value, which is zero ([which means no expiration](https://developer.wordpress.org/reference/functions/wp_cache_set/)).

Should we define an expiration here? WDYT @renatho @aaronfc @alexsanford?

Reply:
humm... that's a good point.

Now I am wondering if we really need any cache here at all. In the end this is just a query (with multiple joins, which I have already given my opinion [here](https://github.com/Automattic/sensei/pull/5294#discussion_r913421183)) but there is no heavy processing anywhere.

My reasoning is:
this could change at any point by a user (assigning new lessons to a course or moving lessons between courses), so this **must not have unlimited expiration time**... 
I would use something low (minutes) to avoid problems with "old-fashioned" information.
An even better alternative would be intercepting the places where the underlying information can be changed and invalidate the cache actively there.
These are the usual issues with caches and I was wondering if we really need the cache at all.

Reply:
hmmm...Yeah, I think the maximum expiration time here would be just a few minutes, maybe 1 or 5 minutes, something like that.

Also, I wouldn't personally leave this without ANY caching at all, just to avoid a possible database overload (even tho I doubt that's possible with such a simple query).

About cache invalidation...that would be waaay too annoying because we would have to guess which cache keys a given lesson is part of, which involves a big amount of tracking, and is just not worth doing at all in this case. I'll update it

Reply:
I think I just discovered a good way to fix that.

I can change the group used for caching here from `sensei` to something like `sensei/temporary` and set that group as non-persistent, using [wp_cache_add_non_persistent_groups](https://developer.wordpress.org/reference/functions/wp_cache_add_non_persistent_groups/).

That way the expiration will still be checked, but we have the warranty (well, sort of, the official implementation is empty because the native object cache is always temporary, so non-persistent) that the data will not be saved to some persistent cache or something, which means that it will only be saved for the duration of the request.

Does that sound good enough, cc @alexsanford @aaronfc? Sounds similar to caching in a variable of some sort, and I guess that would make phpcs happy anyway..

Reply:
Yeah, that sounds like a good solution to me 🤷